libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1411 @@
+                                           keep_only_new_actions,
+                                           progresses);
+}
+
+GHashTable *
+eos_updater_util_filter_for_existing_flatpak_ref_actions (GHashTable *ref_actions,
+                                                          GHashTable *progresses)
+{
+  return filter_flatpak_ref_actions_table (ref_actions,
+                                           keep_only_existing_actions,
+                                           progresses);
+}
+
+const gchar *
+eos_updater_util_pending_flatpak_deployments_state_path (void)
+{
+  return eos_updater_get_envvar_or ("EOS_UPDATER_TEST_UPDATER_FLATPAK_UPGRADE_STATE_DIR",
+                                    LOCALSTATEDIR "/lib/eos-application-tools/flatpak-autoinstall.progress");
+}
+
+const gchar *
+eos_updater_util_flatpak_autoinstall_override_paths (void)
+{
+  return eos_updater_get_envvar_or ("EOS_UPDATER_TEST_UPDATER_FLATPAK_AUTOINSTALL_OVERRIDE_DIRS",
+                                    LOCALSTATEDIR "/lib/eos-application-tools/flatpak-autoinstall.d;"
+                                    SYSCONFDIR "/eos-application-tools/flatpak-autoinstall.d");
+}
+
+GHashTable *
+eos_updater_util_flatpak_ref_action_application_progress_in_state_path (GCancellable  *cancellable,
+                                                                        GError       **error)
+{
+  const gchar *state_file_path = eos_updater_util_pending_flatpak_deployments_state_path ();
+  g_autoptr(GKeyFile) state_key_file = g_key_file_new ();
+  g_auto(GStrv) groups = NULL;
+  g_autoptr(GHashTable) ref_action_progress_for_files = g_hash_table_new_full (g_str_hash,
+                                                                               g_str_equal,
+                                                                               g_free,
+                                                                               NULL);
+
+  GStrv groups_iter = NULL;
+  g_autoptr(GError) local_error = NULL;
+
+  /* Read the key file for sections about the application progress
+   * of each autoinstall file */
+  if (!g_key_file_load_from_file (state_key_file,
+                                  state_file_path,
+                                  G_KEY_FILE_NONE,
+                                  &local_error))
+    {
+      if (!g_error_matches (local_error, G_FILE_ERROR, G_FILE_ERROR_NOENT))
+        {
+          g_propagate_error (error, g_steal_pointer (&local_error));
+          return NULL;
+        }
+
+      return g_steal_pointer (&ref_action_progress_for_files);
+    }
+
+  /* Enumerate each section. The section name is the path to the file */
+  groups_iter = groups = g_key_file_get_groups (state_key_file, NULL);
+
+  while (*groups_iter != NULL)
+    {
+      const gchar *source_path = *groups_iter;
+      g_autofree gchar *progress_string = NULL;
+      guint64 progress;
+
+      /* We need to use g_key_file_get_value here to guard against errors */
```

Comment:
To guard against what kind of errors? Why not `g_key_file_get_integer()`?

Reply:
See https://developer.gnome.org/glib/stable/glib-Key-value-file-parser.html#g-key-file-get-integer - "If key cannot be found then 0 is returned and error is set to G_KEY_FILE_ERROR_KEY_NOT_FOUND. Likewise, if the value associated with key cannot be interpreted as an integer, or is out of range for a gint, then 0 is returned and error is set to G_KEY_FILE_ERROR_INVALID_VALUE.".

That said, we could drop this if we did not allow "0" to be in-range.